### PR TITLE
Add jet shutdown hook, disable imdg one

### PR DESCRIPTION
**Pull request by gurbuzali**
By default, imdg adds its own shutdown hook, which calls either shutdown or terminate according to the shutdown-hook-policy. We've introduced graceful shutdown of jobs but it is only called if explicitly called `JetInstance.shutdown()`. If the processor is killed externally jobs are not shutdown gracefully.

With this PR, we disable the imdg shutdown hook and add our own hook. In the hook, if the policy is graceful then the jobs and underlying imdg instance are shutdown gracefully otherwise we terminate the imdg instance without shutting down the jobs.

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java [hunk 1]
```diff
@@ -67,6 +71,7 @@ public static JetInstance newJetInstance(JetConfig config) {
         configureJetService(config);
         HazelcastInstanceImpl hazelcastInstance = ((HazelcastInstanceProxy)
                 Hazelcast.newHazelcastInstance(config.getHazelcastConfig())).getOriginal();
+        Runtime.getRuntime().addShutdownHook(shutdownHookThread(hazelcastInstance));
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java]: We should also remove the shutdown hook.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java [hunk 2]
```diff
@@ -68,9 +68,8 @@ public static JetInstance newJetInstance(JetConfig config) {
         configureJetService(config);
         HazelcastInstanceImpl hazelcastInstance = ((HazelcastInstanceProxy)
                 Hazelcast.newHazelcastInstance(config.getHazelcastConfig())).getOriginal();
-        JetInstanceImpl jetInstance = new JetInstanceImpl(hazelcastInstance, config);
-        jetInstance.registerShutdownHook();
-        return jetInstance;
+        JetService service = hazelcastInstance.node.nodeEngine.getService(JetService.SERVICE_NAME);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java]: you can also use `Util.getJetInstance`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetInstanceImpl.java [hunk 3]
```diff
@@ -51,6 +51,7 @@ public JetInstanceImpl(HazelcastInstanceImpl hazelcastInstance, JetConfig config
         this.nodeEngine = hazelcastInstance.node.getNodeEngine();
         this.config = config;
         this.shutdownHookThread = shutdownHookThread(nodeEngine);
+        Runtime.getRuntime().addShutdownHook(shutdownHookThread);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetInstanceImpl.java]: the hook shouldn't be created here as JetInstanceImpl is just a wrapper and can be created several times. It should be created in JetService instead.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java [hunk 4]
```diff
@@ -150,5 +151,6 @@ static void configureJetService(JetConfig jetConfig) {
         for (String prop : jetProps.stringPropertyNames()) {
             hzConfig.getProperties().setProperty(prop, jetProps.getProperty(prop));
         }
+        hzConfig.setProperty(GroupProperty.SHUTDOWNHOOK_ENABLED.getName(), "false");
```
[the remark this answers is not in the document]
Comment (gurbuzali) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java]: should we introduce a jet-specific property then ? Or we can check the `SHUTDOWNHOOK_ENABLED` property if it is false, do nothing if it is true add our own hook and set it to false to disable imdg hook

[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/Jet.java]: I think we'll need to disable the original one somehow, and copy the original setting to another property and check that one during startup.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java [hunk 5]
```diff
@@ -51,21 +52,25 @@
...
+import static com.hazelcast.spi.properties.GroupProperty.SHUTDOWNHOOK_POLICY;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
 public class JetService
         implements ManagedService, ConfigurableService<JetConfig>, PacketHandler, MembershipAwareService,
         LiveOperationsTracker {
 
+    public static final String SERVICE_NAME = "hz:impl:jetService";
     public static final int MAX_PARALLEL_ASYNC_OPS = 1000;
+    public static final HazelcastProperty JET_SHUTDOWN_HOOK_ENABLED =
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: can we move this `GroupProperty` ?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java [hunk 6]
```diff
@@ -279,4 +291,15 @@ public void populate(LiveOperations liveOperations) {
         }
         return keys;
     }
+
+    private Thread shutdownHookThread(NodeEngine nodeEngine) {
+        return new Thread(() -> {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: would be good to give the thread a name

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java (hunk 5 above)
Comment (gurbuzali) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: moved it. I was not aware we have jet specific group property class

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java (hunk 6 above)
Comment (gurbuzali) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: "jet.ShutdownThread" ? imdg has "hz.ShutdownThread"

Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JetService.java]: Yes that's fine.

**Pull request merged**